app/src/main/java/com/davidtomas/taskyapp/features/agenda/data/agenda/repository/AgendaRepository.kt

```diff
@@ -0,0 +1,82 @@
+import com.davidtomas.taskyapp.features.agenda.domain.model.AgendaItem
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.async
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.launch
+
+class AgendaRepository(
+    private val agendaService: AgendaService,
+    private val eventLocalSource: EventLocalSource,
+    private val reminderLocalSource: ReminderLocalSource,
+    private val taskLocalSource: TaskLocalSource
+) {
+    suspend fun getAgenda(): Result<List<AgendaItem>, DataError.Network> {
+        return coroutineScope {
+            if (returnAgendaItemsIfLocallySaved() != null) {
+                return@coroutineScope Result.Success(returnAgendaItemsIfLocallySaved()!!)
+            } else {
+                Result.Success(getFromRemoteSaveLocallyAndReturnItems())
+            }
+        }
+    }
+
+    private suspend fun CoroutineScope.returnAgendaItemsIfLocallySaved(): List<AgendaItem>? {
+        val events = async { eventLocalSource.getEvents() }.await().getOrNull()?.first()
+        val tasks = async { taskLocalSource.getTasks() }.await().getOrNull()?.first()
+        val reminders = async { reminderLocalSource.getReminder() }.await().getOrNull()?.first()
+        val listAgendaItem = mutableListOf<AgendaItem>()
+        if (!events.isNullOrEmpty()) events.forEach { listAgendaItem.add(it) }
+        if (!tasks.isNullOrEmpty()) tasks.forEach { listAgendaItem.add(it) }
+        if (!reminders.isNullOrEmpty()) reminders.forEach { listAgendaItem.add(it) }
+
+        return if (!listAgendaItem.isNotEmpty()) listAgendaItem else null
+    }
+
+    private suspend fun CoroutineScope.getFromRemoteSaveLocallyAndReturnItems(): List<AgendaItem> {
+        val listAgendaItem = mutableListOf<AgendaItem>()
+        agendaService.getAgenda().map { agendaResponse ->
+            val job1 = launch {
+                agendaResponse.eventResponses.map { eventResponse ->
+                    val eventModel = eventResponse.toEventModel()
+                    eventLocalSource.saveEvent(eventModel)
+                    listAgendaItem.add(eventModel)
+                    eventModel
+                }
+            }
+            val job2 = launch {
+                agendaResponse.reminderResponses.map { reminderResponse ->
+                    val reminderModel = reminderResponse.toReminderModel()
+                    reminderLocalSource.saveReminder(reminderModel)
+                    listAgendaItem.add(reminderModel)
+                    reminderModel
+                }
+            }
+            val job3 = launch {
+                agendaResponse.taskResponses.map { taskResponse ->
+                    val taskModel = taskResponse.toTaskModel()
+                    taskLocalSource.saveTask(taskModel)
+                    listAgendaItem.add(taskModel)
+                    taskModel
+                }
+            }
```

Comment:
With these launch blocks you're fetching events, tasks and reminders in parallel, but you're still saving each single agenda item sequentially in the corresponding coroutine. You'd either need to expose a function from your DAO which lets you insert a list of items or you also launch one coroutine per `saveEvent` call and so on